item-impl/src/main/java/com/example/auction/item/impl/ItemServiceImpl.java

```diff
@@ -71,67 +74,73 @@ public ItemServiceImpl(PersistentEntityRegistry registry, BiddingService bidding
 
 
     @Override
-    public ServiceCall<Item, Item> createItem() {
+    public ServiceCall<ItemData, Item> createItem() {
         return authenticated(userId -> item -> {
-            if (!userId.equals(item.getCreator())) {
-                throw new Forbidden("User " + userId + " can't created an item on behalf of " + item.getCreator());
-            }
+            // ItemData doesn't contain a creatorId so the Principal _is_ the creator.
             UUID itemId = UUIDs.timeBased();
-            PItem pItem = new PItem(itemId, item.getCreator(), item.getTitle(), item.getDescription(),
-                    item.getCurrencyId(), item.getIncrement(), item.getReservePrice(), item.getAuctionDuration());
-            return entityRef(itemId).ask(new PItemCommand.CreateItem(pItem)).thenApply(done -> convertItem(pItem));
+            PItemData details = Mappers.fromApi(item);
+            // TODO: remove this PItem from here and pass the 3 values in the command?
+            PItem pItem = new PItem(itemId, userId, details);
+            return entityRef(itemId)
+                    .ask(new CreateItem(pItem))
+                    .thenApply(done -> Mappers.toApi(pItem));
         });
     }
 
     @Override
-    public ServiceCall<Item, Done> updateItem(UUID id) {
+    public ServiceCall<ItemData, Item> updateItem(UUID itemId) {
         return authenticated(userId -> item -> {
-            // todo implement
-            return null;
+            // userId may not match the item's creator Id. This is validated at Command Processing in the PersistentEntity
+            PItemData details = Mappers.fromApi(item);
+            PItemCommand.UpdateItem updateItem = new UpdateItem(userId, details);
+            return entityRef(itemId)
+                    .ask(updateItem)
+                    .handle((pitem, updateException) -> {
+                                if (updateException != null) {
+                                    throw new TransportException(TransportErrorCode.fromHttp(409),
```

Comment:
I don't think 400 qualifies. The client did nothing wrong. The events happened in an order that caused the update to fail. When the client sent the request, it might have still been processable but in a millisecond, the event to close the auction happened.

About 422, I think it's reserved for WebDAV.

I think a 409 applies in this case, but the response should include more info so the client can fix the request as requested in the spec: https://tools.ietf.org/html/rfc7231#page-60